### PR TITLE
[Feature] Add MonoFlex Loss

### PR DESCRIPTION
**Pull request by ZCMax**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation
add monoflex loss

## Modification
add monoflex_iou_loss and multibin_loss

## BC-breaking (Optional)
Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)
If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet3d/models/losses/multibin_loss.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        cls_ce_loss = F.cross_entropy(
+            vector_orientations[:, (i * 2):(i * 2 + 2)],
+            gt_orientations[:, i].long(),
+            reduction='mean')
+        # regression loss
+        valid_mask_i = (gt_orientations[:, i] == 1)
+        cls_losses += cls_ce_loss
+        if valid_mask_i.sum() > 0:
+            s = num_dir_bins * 2 + i * 2
+            e = s + 2
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/multibin_loss.py:41]: start and end

File: mmdet3d/models/losses/multibin_loss.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+import torch
+from torch import nn as nn
+from torch.nn import functional as F
+
+from mmdet.models.builder import LOSSES
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def multibin_loss(vector_orientations, gt_orientations, num_dir_bins=4):
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/multibin_loss.py:11]: pred_orientations
gt_orientations

File: mmdet3d/models/losses/monoflex_ious_loss.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+import torch
+from torch import nn as nn
+
+from mmdet.models.builder import LOSSES
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def monoflex_iou_loss(pred, target, mode='iou', eps=1e-7):
+    """FCOS-style IoU loss.
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/monoflex_ious_loss.py:11]: Can we use the iou loss in the mmdet?

Comment (ZCMax) [reply mmdet3d/models/losses/monoflex_ious_loss.py:11]: If we want to use the iou loss in mmdet,  it may need some conversion before feeding into the loss function

File: mmdet3d/models/losses/monoflex_iou_loss.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from torch import nn as nn
+
+from mmdet.models.builder import LOSSES
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def monoflex_iou_loss(pred, target, mode='iou', eps=1e-7):
```
Comment (Tai-Wang) [inline mmdet3d/models/losses/monoflex_iou_loss.py:10]: Can refactor to re-use IoU loss from mmdet

Comment (ZCMax) [reply mmdet3d/models/losses/monoflex_iou_loss.py:10]: I will remove the IoU loss in this PR.

**Review by Tai-Wang — APPROVED** [review]
